zhaquirks/tuya/builder/__init__.py

```diff
@@ -0,0 +1,493 @@
+                    endpoint_id=endpoint_id,
+                )
+            }
+        )
+        self.tuya_data_point_handlers.update({dp_id: dp_handler})
+        return self
+
+    def tuya_dp_attribute(
+        self,
+        dp_id: int,
+        attribute_name: str,
+        ep_attribute: str = TuyaMCUCluster.ep_attribute,
+        converter: Optional[Callable[[Any,], Any,]] = None,
+        dp_converter: Optional[Callable[[Any,], Any,]] = None,
+        endpoint_id: Optional[int] = None,
+        dp_handler: str = "_dp_2_attr_update",
+        type: type = t.uint16_t,
+        access: foundation.ZCLAttributeAccess = foundation.ZCLAttributeAccess.NONE,
+        is_manufacturer_specific=True,
+    ) -> QuirkBuilder:  # fmt: skip
+        """Add an Tuya DataPoint and corresponding AttributeDef."""
+        self.tuya_attribute(
+            dp_id=dp_id,
+            attribute_name=attribute_name,
+            type=type,
+            access=access,
+            is_manufacturer_specific=is_manufacturer_specific,
+        )
+        self.tuya_dp(
+            dp_id=dp_id,
+            ep_attribute=ep_attribute,
+            attribute_name=attribute_name,
+            dp_converter=dp_converter,
+            converter=converter,
+            endpoint_id=endpoint_id,
+            dp_handler=dp_handler,
+        )
+        return self
+
+    def tuya_switch(
+        self,
+        dp_id: int,
+        attribute_name: str = "on_off",
+        endpoint_id: int = 1,
+        force_inverted: bool = False,
+        invert_attribute_name: str | None = None,
+        off_value: int = 0,
+        on_value: int = 1,
+        entity_platform=EntityPlatform.SWITCH,
+        entity_type: EntityType = EntityType.CONFIG,
+        initially_disabled: bool = False,
+        attribute_initialized_from_cache: bool = True,
+        translation_key: str | None = None,
+        fallback_name: str | None = None,
+    ) -> QuirkBuilder:
+        """Add an EntityMetadata containing SwitchMetadata and return self.
+
+        This method allows exposing a switch entity in Home Assistant.
+        """
+        self.tuya_dp_attribute(
+            dp_id=dp_id,
+            attribute_name=attribute_name,
+            type=t.Bool,
+            access=foundation.ZCLAttributeAccess.Read
+            | foundation.ZCLAttributeAccess.Write,
+        )
+        self.switch(
+            attribute_name=attribute_name,
+            cluster_id=TUYA_CLUSTER_ID,
+            endpoint_id=endpoint_id,
+            force_inverted=force_inverted,
+            invert_attribute_name=invert_attribute_name,
+            off_value=off_value,
+            on_value=on_value,
+            entity_platform=entity_platform,
+            entity_type=entity_type,
+            initially_disabled=initially_disabled,
+            attribute_initialized_from_cache=attribute_initialized_from_cache,
+            translation_key=translation_key,
+            fallback_name=fallback_name,
+        )
```

Comment:
Not sure we want to do this, but I think we could replace all default arguments (passed down to the normal `.switch()` method with `*args, **kwargs`. Then, just pass those in.
It should contain all arguments that we did not explicitly define in our method definition.

Reply:
I have no strong preference on it, I had started with *args and **kwargs, but looking through the repo it looked like that wasn't the typical pattern, so I went the other way. Happy to change it, if that's the consensus.

Reply:
Personally, I think I'd rather go with `**kwargs` for all arguments just passed to the underlying v2 builder methods.

Reply:
One issue with **kwargs would we lose type hinting. Might be worth adding appropriate TypedDicts for each function in zigpy and then importing them here.

Reply:
So I think we need to change Zigpy to do this cleanly. We would have to add **kwargs to the method signatures and then here we would use **locals() to call the methods or do something like this hack…

```
import inspect

def foo(bar=baz, test=blah, **kwargs):
    # Gather all local variables (including bar, test, and kwargs)
    all_args = {**locals(), **kwargs}

    # Get the signature of the inner function (another_method)
    sig = inspect.signature(another_method)

    # Filter all_args to include only parameters that are explicitly defined in another_method
    filtered_args = {k: v for k, v in all_args.items() if k in sig.parameters}

    # Pass only filtered arguments to another method
    another_method(**filtered_args)

def another_method(bar, test):
    print(f"bar: {bar}, test: {test}")

# Example usage:
foo(bar=1, test=2, extra_arg="extra")
```

why don’t we leave it as is for now and we can address it later in a streamlining PR? (Unless I am misunderstanding or missing something)

Reply:
Ok, let's leave it how it is at the moment. We can reconsider this in the future.